### ISSUE TITLE
Improve UI redraws

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#719]
Currently the whole UI is redrawn when a widget changes (becomes "dirty"). It might(!) be more effective if only the "dirty" widgets are redrawn. 

This would also allow the (active) widgets to animate themselves (e.g. animated cursor, scrolling labels when there is not enough space or some other fancy stuff ), since they can be queried in regular 60 Hz intervals then.

The code changes required should be quite small, but the devil is in the detail.
- add `isDirty` method to each `GuiObject`
- change `setDirty` to update an internal state instead of informing the dialog
- when rendering a dialog, call `isDirty` for all its widgets 
- if a dirty widget is found either redraw the whole dialog or only the dirty widget(s) and reset their dirty state (whatever is more efficient)

**Comment by sa666666** [comment]
Right now, any change to a dialog (which has an `FBSurface` behind the scenes) requires redrawing the whole dialog.  This is because it's just as fast to redraw a surface than it is to only redraw parts of it.  For OpenGL, it has to do with calling `glTexImage2D` _once_ on the whole surface, vs. (potentially multiple) calls to `glTexSubImage2D` on many parts of the surface.  Direct3D has something similar.

So, if we conclude that we'll only ever draw the entire dialog rather than parts of it, then it makes no sense for each widget to track its own dirty state.  If _it_ is dirty, then by extension the whole surface is dirty, and it can just inform the dialog of that fact.  And that's exactly what the current code does.  Widgets set a flag in `Dialog` that the dialog needs to be redrawn.  Turning this logic around, and having the dialog query each widget for its 'dirtiness' is slower, and in the end results in the exact same thing; the entire dialog is redrawn.

For the debugger, it happens that the entire thing is a dialog (aka, `DebuggerDialog`), so redraws happen on the entire thing.

Long story short; if in the end a widget being dirty implies a dialog needs to be redrawn, then the code already works that way, and we gain absolutely nothing by changing how it works.

As for regular querying, this happens in `DialogContainer::updateTime`.  At regular intervals, 'time' is updated in that class, which then can pass it on to its constituent dialogs (and from there to its constituent widgets).  That's how we currently check whether a mouse click is still being held down.  It should be possible to add code to this class to send 'time' info to dialogs/widgets (only if they register that they want it, though!) from this class.  And that would involve not having to touch the rendering subsystem at all.

**Comment by sa666666** [comment]
Also, whether something is dirty and whether it should animate are two different things.  Even if we reworked the widgets to have their own dirty state, they'd still need to know _when_ to animate.  Hence they need _time_ information, as mentioned above.  Once they have that, it doesn't matter if the widget tells the dialog, or the dialog queries the widget; the redrawn needs to happen because _it's time_ has come to happen.

**Comment by thrust26** [comment]
> Right now, any change to a dialog (which has an FBSurface behind the scenes) requires redrawing the whole dialog. This is because it's just as fast to redraw a surface than it is to only redraw parts of it. 

OK for bringing the surface to display. But the actual `drawWidget` methods require CPU time too.

**Comment by sa666666** [comment]
Right; it's a toss-up.  We would need to test.  On older systems, the overhead of doing multiple calls to upload parts of a surface/texture was much greater than any drawing done locally, into the buffer.  Maybe that's no longer true.

But the other point still stands.  Even if we change the code to work that way, it still doesn't help with animations unless the widget has access to 'time' info.  And the dialog already has access to that info, so it should be able to pass it to the widget.

**Comment by thrust26** [comment]
>  On older systems, the overhead of doing multiple calls to upload parts of a surface/texture was much greater than any drawing done locally, into the buffer. Maybe that's no longer true.

We could draw the widget individually and then still update the whole surface. 

> But the other point still stands. Even if we change the code to work that way, it still doesn't help with animations unless the widget has access to 'time' info. And the dialog already has access to that info, so it should be able to pass it to the widget.

The widget would have access to time info, because it would be polled in regular intervals if it is dirty. So it can tick its own internal timer.

**Comment by sa666666** [comment]
OK, you've convinced me.  But I think if you want to attempt this, please do it in a branch first.  It has the potential to break things, and will require a lot of testing.  Or maybe I'm just overestimating how complicated it will be again :smiling_imp:

**Comment by thrust26** [comment]
This is definitely worth a branch.

**Comment by thrust26** [comment]
This might also allow mouse overs.

**Comment by thrust26** [comment]
Works quite well already.

Known bugs:
- `RomInfoWidget`drawn at left in `LauncherDialog` if first selected game has no screenshot
- surface cleared in `DebuggerDialog` if other dialog is in front
- some superfluous draws have to be analyzed

**Comment by sa666666** [comment]
Good work so far.  It's not surprising there are some issues.  The old code was tuned/tested for quite some time, so we will have to do this again.  And undoubtedly there will be some assumptions that are no longer valid.  But it will make drawing use less CPU, so that's great.

**Comment by thrust26** [comment]
How complicated would it be to give each dialog its own surface? Then we can save a lot of redraw when closing a dialog or context menu.

**Comment by sa666666** [comment]
It already has it.  This is in the private instance variable list for `Dialog`:
```    shared_ptr<FBSurface> _surface;
```

**Comment by sa666666** [comment]
There are also still issues with double-buffering, and `PopupWidget` in particular.  Drawing the 'menu' sometimes doesn't show the underlying dialog.

Also, switching to fullscreen sometimes has garbage outside the dialog area (probably missing a `clear()` somewhere).  Some platforms don't zero video memory, so we have to take care of it.

**Comment by thrust26** [comment]
I see now too. Instead of drawing all the bottom dialogs when closing the top one, rendering their surfaces should do. The current code couples both actions.

I have experimented with that now, but I did not notice the problems your described. 

BTW: Do you know how I can darken a surface? The scanlines seem to do that, no? I suppose the process is reversible?

**Comment by sa666666** [comment]
> I see now too. Instead of drawing all the bottom dialogs when closing the top one, rendering their surfaces should do. The current code couples both actions.

Yes, the code couples drawing and rendering.  Re-rendering is all that's needed; redrawing only needs to happen when a widget itself changes.  This is a holdover from software rendering, where there were no surfaces, and a dirty widget meant a redraw was always required.  There was no concept of 'render surfaces' at all.

> I have experimented with that now, but I did not notice the problems your described.

Yes, some of these things only happen in Windows, some in Linux, and some different ones in Mac.  That's why I mentioned that much testing will be required.  And it may also explain some of the 'strangeness' of the current code; the behaviour of video drivers isn't consistent on different platforms.

> BTW: Do you know how I can darken a surface? The scanlines seem to do that, no? I suppose the process is reversible?

What do you mean by darken?  If like scanlines, then it's just a texture with alpha set to a certain level.  Also not sure what you mean by reversible.  Just re-render with a different alpha level, I guess.

**Comment by thrust26** [comment]
I will commit my changes regarding surfaces (today or tomorrow) and then we can start testing.

**Comment by thrust26** [comment]
With "darken" I mean that dialogs in the background should become darker and back to normal brightness when they are on top again. Not sure if I will keep it like that, current Windows only changes the color of the title bar and blinks the title of the active dialog when you click an inactive dialog. But I don't like that very much, I want to know **before** I click wrong. :smile:

Anyway, just to be sure:  For darkening, I create a new surface per dialog and render it on top of that dialog.

**Comment by sa666666** [comment]
> Anyway, just to be sure: For darkening, I create a new surface per dialog and render it on top of that dialog.

Yes, that's basically how scanlines work:
```

  // Generate scanline data, and a pre-defined scanline surface
  constexpr uInt32 scanHeight = TIAConstants::frameBufferHeight * 2;
  std::array<uInt32, scanHeight> scanData;
  for(uInt32 i = 0; i < scanHeight; i += 2)
  {
    scanData[i]   = 0x00000000;
    scanData[i+1] = 0xff000000;
  }
  mySLineSurface = myFB.allocateSurface(1, scanHeight, interpolationModeFromSettings(myOSystem.settings()), scanData.data());
```
I suppose if you want to get fancy, you could just create a 1x1 pixel surface filled with 0xFF, and then render it on top of the dialog with the appropriate alpha level.  Note that in the above code, the scanline texture is only 1 pixel wide.  It is stretched horizontally, so we save memory by not having to allocate the entire width.  It is 'stretched' over the surface as it's rendered, and this is essentially free (hardware scaling).

And to be even more fancy, you only really need _one_ such surface, which can be shared by _all_ dialogs.  It can be a static class variable that is shared amongst all dialogs.

**Comment by thrust26** [comment]
Thanks.

Cool idea with the single, stretched surface. That explains why the scanline surface has only one dimension.

**Comment by sa666666** [comment]
You're not the only one that likes to optimize things :smiley:

**Comment by thrust26** [comment]
@sa666666 I optimized that much, that now when I close a dialog (`DialogContainer::removeDialog()`), I don't have to draw anything again. But now the screen is not updated until the next draw happens. And I haven't found out to enforce an update.

**Comment by sa666666** [comment]
I guess the only way to do it is have a full refresh on every state change (when a dialog is opened or closed, toggling windowed/fullscreen mode, etc).  It may not be 100% efficient, but at least we will see the output :smile: 

I think this is probably related to the 'garbage' I see when switching to fullscreen mode.  A `clear()` is missing, so it's showing the old buffer contents.

**Comment by sa666666** [comment]
I'm pretty sure that's why the old code re-draws the entire dialog stack when a dialog is closed.

**Comment by sa666666** [comment]
`FrameBuffer::update(true)` is supposed to enforce a full redraw or re-render, no matter if things are dirty or not.  But during refactoring, you may have changed the intent here??

**Comment by thrust26** [comment]
Yes, that was changed intentionally. There are no pending draws, just the dialog surfaces were rendered again. Now I want to make this visible.

**Comment by sa666666** [comment]
You still need to call `FBSurfaceSDL2::render()` for all surfaces.  The method `FBBackendSDL2::renderToScreen()` just finalizes all the renders up to that point, and 'pushes' them all out to the video hardware.  So if you haven't called `render` at some point, `renderToScreen` won't actually do anything, I think.  The code is:
```
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void FBBackendSDL2::renderToScreen()
{
  ASSERT_MAIN_THREAD;

  // Show all changes made to the renderer
  SDL_RenderPresent(myRenderer);
}
```
Note the 'show all changes' comment.  Without calling render on surfaces, there are no changes, so nothing happens.

Not sure if this is what the problem is.  But it's a place to start.

**Comment by thrust26** [comment]
Well, if you look at `DialogContainer::removeDialog()`I **am** rendering all surfaces again.

**Comment by sa666666** [comment]
Sorry, I am going only by your comments; I haven't had time to look at the code.  Maybe later this evening.

**Comment by sa666666** [comment]
On line 179 in `DialogContainer::removeDialog()`, the `setDirty()` part is commented out.  I understand why this is; to eliminate spurious redraws.  However, adding it back allows the dialog to 'disappear' as it's supposed to.

So there is still some interaction between drawing and rendering; they are not separated yet.  So doing a draw and then a render is working, but a render by itself isn't.  So now we just need to track down what the difference is.

**Comment by sa666666** [comment]
Or, maybe we just have to accept that a full redraw and render is required when closing a dialog.  It's not _that_ inefficient, after all.

**Comment by thrust26** [comment]
Yes, it seems that fully not changed surfaces may not be rendered at all. But I found nothing in the documentation. However, there must be a way, I don't give up easily.

**Comment by sa666666** [comment]
There are also cases where a redraw is not happening when it should, resulting in garbage data in the buffer.  For example:
```
void DialogContainer::draw(bool full)
{
  if(myDialogStack.empty())
    return;

  cerr << "draw " << full << " " << typeid(*this).name() << endl;

  // Make the top dialog dirty if a full redraw is requested
  if(full)
    myDialogStack.top()->setDirty();

  // Draw and render all dirty dialogs
  myDialogStack.applyAll([&](Dialog*& d) {
    if(d->needsRedraw())
      d->redraw();
  });
}
```
Here, the semantics of the draw, when `full` is true, is supposed to be 'force a full redraw'.  In that case, the draw shouldn't depend on `needsRedraw()`, but also on `full`.  So perhaps something like:
```
void DialogContainer::draw(bool full)
{
  if(myDialogStack.empty())
    return;

  cerr << "draw " << full << " " << typeid(*this).name() << endl;

  // Draw and render all dirty dialogs
  myDialogStack.applyAll([&](Dialog*& d) {
    if(full)
      d->setDirty();
    if(d->needsRedraw())
      d->redraw();
  });
}
```

**Comment by thrust26** [comment]
> There are also cases where a redraw is not happening when it should

Can you give me an example?

**Comment by sa666666** [comment]
I can trigger it pretty consistently, but don't know how to explain it.  A video would be better.  Perhaps I can make a video on my camera?

Also, and you're probably not seeing this fact yet; this stuff works differently depending on the backend.  So what's working for you in Direct3D may not work for me in OpenGL.  And I tested on Mac using Metal, and it seems different in some cases too.  That's why I initially had the redraws happening with the renders all the time; the behaviour doesn't seem to be consistent on all platforms.

**Comment by thrust26** [comment]
Yes, the various renderers and platforms need more testing.

Anyway, I tried something again, and suddenly it works!

```
void DialogContainer::removeDialog()
{
  if(!myDialogStack.empty())
  {
    myDialogStack.pop();

    if(!myDialogStack.empty())
    {
      // Rerender all dialogs (TODO: top dialog is rendered twice)
      myDialogStack.applyAll([&](Dialog*& d){
          //d->setDirty();
          d->render();
      });
      myOSystem.frameBuffer().renderToScreen(); // <- this!
    }
  }
}
```

**Comment by thrust26** [comment]
> I can trigger it pretty consistently, but don't know how to explain it. A video would be better. Perhaps I can make a video on my camera?

Sure.

**Comment by sa666666** [comment]
![Screenshot_20201112_192238](https://user-images.githubusercontent.com/16061066/99006132-8c69fb80-251c-11eb-85c4-5e37927f2053.png)
This happened after exiting Game Info dialog, Alt-tabbing away from the window, then back again.  It seems the launcher is drawn, then the dialog.  Then the dialog again.  And finally the launcher draw is called, but it isn't dirty, so nothing happens.

I've run into something like this before.  Basically, there is double-buffering going on here, and if you draw an underlying menu/dialog into one buffer and then switch to another to draw the top-level dialog, the bottom-level one may not be drawn.  Sometimes it gives a black screen (like here), sometimes garbage.  But the reason is that you've swapped the buffer without drawing in it.  Every time `renderToScreen()` is called, it swaps to a new buffer.  And if that buffer isn't fully redrawn, then we get this.  That's why I was drawing all the dialogs each time any one of them changed.  I never did figure out a way around that.

**Comment by sa666666** [comment]
> Yes, the various renderers and platforms need more testing.
> 
> Anyway, I tried something again, and suddenly it works!
> 
> ```
> void DialogContainer::removeDialog()
> {
>   if(!myDialogStack.empty())
>   {
>     myDialogStack.pop();
> 
>     if(!myDialogStack.empty())
>     {
>       // Rerender all dialogs (TODO: top dialog is rendered twice)
>       myDialogStack.applyAll([&](Dialog*& d){
>           //d->setDirty();
>           d->render();
>       });
>       myOSystem.frameBuffer().renderToScreen(); // <- this!
>     }
>   }
> }
> ```

That `renderToScreen()` call is happening at the end of `FrameBuffer::update()`.  If it isn't being called there, it means the change in state in `DialogContainer::removeDialog()` isn't being properly communicated back to the `update()` call.

**Comment by thrust26** [comment]
Yes, that's something I had seen before. But I am optimistic.

BTW: I found that OpenGL has problems with the shading surface. It goes on an off when you tab through the `OptionDialog`  (or move the mouse). Maybe something similar as you described.

**Comment by thrust26** [comment]
> That renderToScreen() call is happening at the end of FrameBuffer::update(). If it isn't being called there, it means the change in state in DialogContainer::removeDialog() isn't being properly communicated back to the update() call.

Not sure what you want to tell me here. What would the `update()` have to know?

**Comment by sa666666** [comment]
Basically, at the end of `FrameBuffer::update()`, it tests whether a change has really happened, and swaps to a new buffer (that's what `renderToScreen` does).  I guess it made sense in the original design to have only the `FrameBuffer::update()` swapping buffers, after all drawing/rendering is done.  Having it happen in other classes seems to be exposing an implementation detail of now the `FrameBuffer` class works (that's the reason why that functionality isn't public).  Perhaps a way to tell `FrameBuffer` to invalidate and update is better?

**Comment by thrust26** [comment]
> Perhaps a way to tell FrameBuffer to invalidate and update is better?

Just what I said some seconds before. :smile:

**Comment by sa666666** [comment]
OK, goodnight.

BTW, I think there was a `FrameBuffer::invalidate()` at some point.  Not sure if it was removed in my rework or yours.  We need something like that again, I guess.

**Comment by thrust26** [comment]
One more thing: OpenGL seems to randomize the surfaces if you put them on screen without rendering them all again. Something to test tomorrow.

**Comment by sa666666** [comment]
Here's a video of switching to fullscreen in emulation mode, then opening the debugger, and finally the RAM search dialog (so the cursor will tick).  Ignore the small size of the UI; I've intentionally turned off HiDPI mode:

[MVI_0858.zip](https://github.com/stella-emu/stella/files/5533925/MVI_0858.zip)

Watch is swap between emulation in background, debugger windows shaded and then not.  The upper dialog is being drawn correctly; everything else isn't.  I fear we're going down a rabbit hole here :smiling_imp: 

Mac does something similar, except the background is often red instead of black.

**Comment by sa666666** [comment]
Actually, I'm seeing similar behaviour in Linux, Windows and Mac.  And this is all different hardware and video cards.  And different drivers (Direct3D, OpenGL, Metal).  So at least we are consistent 😈  But this is actually a good thing, since it indicates a problem in the code, and not something platform-specific.  The latter is always so hard to track down.  So perhaps when we fix the issues on one system, it will be taken care of on another too.

BTW, not to sound discouraging, but this is why I avoided the UI code after working on it for so long.  I really started to get burnt out.

**Comment by sa666666** [comment]
I think one of the main issues is that we must not call `renderToScreen` unless something has actually changed.  It will swap buffers, and the new one won't necessarily match the old one.  And we can't depend on something rendered to one of them being present in the other.

I believe the way the old code worked is that if **anything** changed, the entire dialog stack had to be redrawn.  Now that drawing is separated from rendering, it applies for the latter only:
- a dirty widget draws only that part into the dialog surface texture, which then implies that a render is needed (to flush the changes to the screen)
- a re-render of a top-level dialog (whether its surface has changed or not) implies **all the bottom ones** need to be re-rendered too, even though they haven't changed (this is because if you re-render a top-level surface and not the bottom ones and then swap buffers, the other buffer won't have the underlying dialogs present)
- if any re-rendering has happened in `FrameBuffer::update()`, then the changes are flushed and the buffers swapped with `renderToScreen`.  Conversely, if nothing has changed, then the buffer shouldn't be swapped.

This is what we're seeing here.  Top-level dialogs are being drawn correctly, and underlying ones are not.  That's why we're getting garbage, old contents, red/black colours, etc.  When we re-render, we have to re-render **everything**.  And we should only call `renderToScreen` once per framebuffer update, and only if re-rendering has occurred.

@thrust26, I know this seems slightly less efficient, but I think it's the only way it can work.  At least we've separated drawing from rendering, and only do the former when necessary.  But the latter has to happen on any change, and it has to happen to everything on the screen.

TL;DR:  All the re-renders have to happen together, in the same update, and be pushed to the screen at the end of that update.  Whether the surfaces have actually changed (by widgets drawing into them) is irrelevant.  Dirty surfaces imply a re-render is required.  Double-buffering requires all re-renders to be done together, even ones underneath.

**Comment by thrust26** [comment]
> When we re-render, we have to re-render everything

Yup, that was my point too. Fingers crossed, I will try now.

**Comment by thrust26** [comment]
@sa666666 I have changed the rendering to what you suggested. This now looks fine with all 4 renderers on I have. 

Note: `renderToScreen` is still called from `removeDialog`, as I want to become aware of all potential errors.

**Comment by sa666666** [comment]
I'm in a lab now, and will be for much of today.  So I'm not going to have a lot of time to test this until this evening (10 hours from now).

At preliminary first glance, testing on Mac, all the re-render issues are gone. 👍   So that's great.  I still need to test this in Linux, but I suspect it will be the same.

Now some bugs:
- Switching to fullscreen mode leaves garbage in the background sometimes; this is somewhat related, in that we need a `clear()`.  In fact, it might be better to just do a `clear()` whenever a sequence of re-renders happen, to be sure the buffer is clean.
- On Mac at least, Stella crashes on lines 900-901 of `FrameBuffer::setUIPalette()`:
```
  if(&myOSystem.eventHandler())
    update(true);
```
What is the purpose of this code?  Perhaps I can rework it so it doesn't crash, _if_ I know what the intent is.

**Comment by thrust26** [comment]
That crash comes by no surprise. Actually I was surprised that it works in Windows.

The propose is to force a full UI redraw when the palette has been changed. But initially the code is called without an `EventHandler` existing, so I had to add the check.

I'll look into the fullscreen garbage.

**Comment by sa666666** [comment]
I'm going to comment that part out for now, until we find another solution.

**Comment by thrust26** [comment]
OK. 

I cannot reproduce the fullscreen problem, so I need you help here. The code in `applyVideoMode` does `resetSurfaces` (they should be clear then, no?). And then it executes `update(true)`. So that looks fine. But the garbage indicates that either a surface is not cleared or that it is not rendered again.

BTW: Does the garbage really show when switching **to** fullscreen mode? In the launcher or the emulation or both?

**Comment by sa666666** [comment]
I don't have a lot of time right now, so I can't get into it too deeply.  But it's happening when the dialog (say, the debugger) is smaller than the emulation output.
- start Stella in windowed mode
- start a game
- switch to fullscreen
- switch to the debugger

At that point, if the debugger window is smaller than the TIA output, you'll see that in the background, behind the debugger.  Basically, the debugger is only re-rendering to its own dimensions, but there TIA dimensions are larger.  Or more to the point, in fullscreen mode the UI can be smaller than the 'window' size.  This can't happen in windowed mode, which is why you wouldn't see it there.

So it's another symptom of the same problem; when you do a re-render, you need to start from a clean buffer.  That means a `clear()` is needed at the beginning of the re-render stage.

**Comment by thrust26** [comment]
Ok, now I can reproduce and fix. Thanks!

**Comment by sa666666** [comment]
Fix applies to Windows, but unfortunately not to Linux and Mac.  I have an idea how to fix it, but will have to leave it until later this evening.  I've seen this problem before.

Now more bugs :smiling_imp: 
- Linux needs the full redraw in `EventHandler::handleSystemEvent` 
```
    case SystemEvent::WINDOW_EXPOSED:
    case SystemEvent::WINDOW_RESIZED:
      //myOSystem.frameBuffer().update(true); // force full update
      // TODO: test and maybe force a render update instead
      myOSystem.frameBuffer().update();
      break;
```
Without that full refresh, tabbing away and back again leaves the window completely blank (white).  I think this is related to OpenGL losing all state when the window goes out of focus, and needing a full redraw.  I don't think Direct3D works like that, so that's why you're not seeing it there.

- This next one is a long-standing issue, not caused by the current work (but related to it now).  When exiting a `PopupWidget`  (aka, a `ContextMenu`) with Escape, it seems that the underlying dialog isn't receiving focus events.  So if you click on such a widget, escape from it, then move the mouse around, none of the other widgets are highlighted.  If you click in the dialog and move again, _then_ it works as expected.  This is happening on all ports, so it's not a driver thing.

**Comment by thrust26** [comment]
> I think this is related to OpenGL losing all state when the window goes out of focus, and needing a full redraw. I don't think Direct3D works like that, so that's why you're not seeing it there.

I have OpenGL avail for testing too, there it doesn't happen. But I think a full render should do, no?

**Comment by sa666666** [comment]
When I put `true` in the brackets, it works fine.  So basically when we revert the code to what was there before, it works fine.

**Comment by thrust26** [comment]
Yes, that's a full redraw and render. I plan to change the interface to force full render too.

**Comment by sa666666** [comment]
Also, I'm not sure how OpenGL works in Windows is how it works everywhere else.  So it's not always a good test base for that.

> Yes, that's a full redraw and render. I plan to change the interface to allow forcing full render only too.

I will test when you have that complete.  Perhaps a full re-render is all that's needed.

**Comment by sa666666** [comment]
Hmm, an awful lot of comments for a prio3 issue 😄   Some more bugs.

First one.  Select a breakpoint in the UI, then move the mouse back and forth over the area (I guess to lose and gain focus).  The BP 'circle' flashes, indicating it's not being redrawn in all frames.  You just recently added this code, so perhaps you're still working on it.

Second one:
- Go to ROM launcher
- Select a ROM (left click, don't start; just select)
- Open the options dialog
- Close the options dialog
- Click back in the ROM launcher list widget, on the ROM that was previously selected
- The ROM still has a 'rectangle' around it, as if focus isn't returned to that `ListWidget`
- Move mouse around, over the scrollbar area.  _Now_ you see the draw happen correctly.

I can provide videos to illustrate these issues, if you like.

**Comment by thrust26** [comment]
You mean that the ROM should be displayed inverted again, after clicking after returning from Options, correct?

**Comment by sa666666** [comment]
Maybe a video helps here.  Sorry for the quality and shaking.
[MVI_0860.zip](https://github.com/stella-emu/stella/files/5541178/MVI_0860.zip)

**Comment by thrust26** [comment]
Yup, there is either a redraw missing or the selection shouldn't be removed when you open another dialog. Or both. :smile:

**Comment by thrust26** [comment]
The breakpoints issues come from the very odd, unique `RomListWidget` implementation. It stacks `Checkbox` widgets on top of itself.

**Comment by sa666666** [comment]
> Yup, there is either a redraw missing or the selection shouldn't be removed when you open another dialog. Or both.

Pretty sure the selection should be removed, as that's happening when you click the 'Options' button, and focus switches from the list to the button.  So that part is working.  It's after the dialog closes and you re-click on the list that the redraw (or re-render) isn't happening.  Interestingly, if you click on _another_ item in the list, the update is done correctly.  So it's only when clicking on the original item that the redraw doesn't happen.  Maybe something in the `ListWidget` code is checking if what you clicked is already selected, and not issuing a refresh in that case.

> The breakpoint issues come from the very odd, unique `RomListWidget` implementation. It stacks `CheckboxWidgets` on top of itself.

It's not surprising that in touching all these different parts of the code, you're encountering some 'strangeness'.  That's what happens when you dig up skeletons :smirk:

**Comment by thrust26** [comment]
I changed the code to that receiving or loosing focus doesn't issue a redraw when the state doesn't change. And somehow the widget did not loose focus.

**Comment by sa666666** [comment]
Whatever the reason, I feel the old behaviour is indeed correct.  When you click away from the list to the button, the button gains focus.  That means the list should lose it.  And _that_ means the currently selected item in the list should be drawn with an outlined rectangle rather than a solid one.

Similarly, when clicking back to the list, the currently selected item should regain focus, and change from an outline back to a solid rectangle.  I think however we fix the drawing issues, this should be the focus behaviour we aim for.

**Comment by thrust26** [comment]
This widget doesn't only use the dotted frame for focus display but also changes the display of the currently selected item.  But when a widget regains focus, it only updates the dotted frame. However tabs work as required. So there is more...

**Comment by sa666666** [comment]
I guess we never noticed this before, since it was redrawing the entire thing every time.

**Comment by thrust26** [comment]
Yup, the separation between drawing and renderings reveals coding which doesn't follow the general pattern. I hope there are not too many of these. I do not like hacking into a clean approach,

BTW: Is the fullscreen garbage is solved on your side too?

**Comment by sa666666** [comment]
> Yup, the separation between drawing and renderings reveals coding which doesn't follow the general pattern. I hope there are not too many of these. I do not like hacking into a clean approach,

This was to be expected, I think.  The changes you're doing are pretty invasive, design-wise.  I'm surprised it didn't reveal more issues.

> BTW: Is the fullscreen garbage is solved on your side too?

Unfortunately, it's still happening on Linux and Mac, but not Windows.  So I will have to fix this one, since you can't replicate it on your end.  Hope to look at it this evening.

**Comment by thrust26** [comment]
> Unfortunately, it's still happening on Linux and Mac, but not Windows. So I will have to fix this one, since you can't replicate it on your end. Hope to look at it this evening.

A full redraw should do, but I wonder why this is necessary.

**Comment by sa666666** [comment]
> Maybe a `clear()` in `update()` would do in case of rendering?

Pretty sure that's the way it was before we started refactoring.

**Comment by thrust26** [comment]
Yea, not the best idea I had lately. :smile: 

Next try: The garbage we see in launcher and debugger are definitely from the TIA surface. So clearing that one when leaving emulation has to be ensured.

**Comment by sa666666** [comment]
No, not necessarily.  On a Mac I see a red screen, and on Linux I sometimes see a mixture of many different things; whatever was in video RAM at that time.  So it's not tied strictly to emulation mode or the TIA surface.

Let me play with it a little.  I've seen this before, and I _did_ fix it eventually (even though I don't remember how right now).

**Comment by thrust26** [comment]
That sounds like uninitialized surfaces. Or what else could be displayed there?

**Comment by sa666666** [comment]
It's due to not clearing the entire surface.  In the current code, TIA emulation or dialogs mostly don't take up the entire fullscreen area; they're centered in the middle.  When we do the clear, it has to clear the entire surface, not just the surface of any particular dialog.

When I said it's not strictly related to TIA emulation, I meant I can get it to happen without even starting a ROM.  So just adding code around the TIA surface functionality won't fix it; it has to happen in `update`.

It's definitely a dirty, uninitialized buffer causing the problem.  The fix is to issue a `clear` before rendering.  The main issue is, where to place the `clear` so it only happens when necessary.

**Comment by sa666666** [comment]
Also, some rendering backends don't zero-out buffers, while others do.  I suspect that's why you're not seeing it in Windows, and I am in Linux/Mac.

**Comment by thrust26** [comment]
How about adding a black background surface?

**Comment by sa666666** [comment]
I suspect a `clear` is faster.  It's typically a single operation in different backends (for OpenGL it's `glClear()`), vs. uploading a texture (`glTexImage2D`).  I forget the operations for Direct3D, but it's the same idea.  Besides, we still need to figure out _where_ to do the operation.  And once we do that, we may as well just use `clear`.

**Comment by thrust26** [comment]
Are we doing double buffering in UI mode too?

**Comment by sa666666** [comment]
Double-buffering is always happening, as that's the way the video mode is opened in SDL.  In fact, all current video backends default to that.  This is the exact reason why we're seeing the 'flashing'; data is in one buffer but not the other, and when calling `renderToScreen`, we're swapping buffers.

Anyway, I want to go away and work on it for a little while now.  Hopefully I'll have some progress for tomorrow.

**Comment by thrust26** [comment]
Sure, don't get distracted by me.

**Comment by thrust26** [comment]
A suggestion:
```
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void DialogContainer::render()
{
  if(myDialogStack.empty())
    return;

  switch(myOSystem.eventHandler().state()) ; <- new!
  {
    case EventHandlerState::LAUNCHER:
    case EventHandlerState::DEBUGGER:
      myOSystem.frameBuffer().clear();
      break;

    default:
      break;
  }
  ...
```

**Comment by sa666666** [comment]
Yep, it seems that fixed it :+1:  I changed the code slightly, to use a method from `EventHandler`, but otherwise it's fine.
```
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void DialogContainer::render()
{
  if(myDialogStack.empty())
    return;

  cerr << "full re-render " << typeid(*this).name() << endl;

  // Make sure we start in a clean state (with zero'ed buffers)
  if(!myOSystem.eventHandler().inTIAMode())
    myOSystem.frameBuffer().clear();

  // Render all dialogs
  myDialogStack.applyAll([&](Dialog*& d) {
    d->render();
  });
}
```

**Comment by sa666666** [comment]
Unfortunately found another bug :bug: Changing the sliders in the Video/Audio palette tab no longer changes the actual colour swatches.  The TIA output is updated; it's just the palette squares themselves that aren't changing.

**Comment by sa666666** [comment]
One bug fixed (garbage in background in fullscreen), another one found :bug: .  List above has been updated with progress so far.

There are several ways to trigger this; I will outline only one:
- Open a ROM
- Open options dialog (tab key)
- Exit options dialog
- Resize TIA window (increase or decrease)
- Open options dialog again
- For a moment, the dialog is drawn in a different position, then 'snaps' to the center

So basically, it looks like a render is happening with the dialog from its previous location, then a redraw/re-render positions and draws it in the correct location.  Not a terribly huge bug, but it looks kind of ugly.

**Comment by thrust26** [comment]
I suppose the same happens with `CommandMenu` too, correct?

**Comment by thrust26** [comment]
> exit from PopupWidget (ContextMenu) with Escape requires a mouse click on dialog for events/redraws to happen

Grrr, can't reproduce again. You mean the `ContextMenu` is still displayed? Until an event happens?

**Comment by sa666666** [comment]
I will post a video soon.

**Comment by sa666666** [comment]
> I suppose the same happens with `CommandMenu` too (I can't reproduce again)?

No!  It only happens for Options, not for CommandMenu.  See video:
[menu_reposition.zip](https://github.com/stella-emu/stella/files/5542541/menu_reposition.zip)

Skip to the 37 second mark to see it in action.  After resizing the TIA window and pressing Tab, the options menu menu jumps for a moment.  I've actually narrowed it down to an interaction between the message gauge at the bottom of the window being onscreen at the same time.  If you wait for the gauge to disappear and then open Options menu, it opens as normal.  Only when you open it _while_ a message box/gauge is onscreen do we get this repositioning.  For some reason this doesn't happen with `CommandMenu`, so it must differ from `OptionsDialog` somehow.

> > exit from PopupWidget (ContextMenu) with Escape requires a mouse click on dialog for events/redraws to happen
> 
> Grrr, can't reproduce again. You mean the `ContextMenu` is still displayed? Until an event happens?

OK, see this video:
[click_after_context_required.zip](https://github.com/stella-emu/stella/files/5542560/click_after_context_required.zip)

Here, I first hover over a few widgets, to show the highlight is working.  Then open a `ContextMenu` (aka, `PopupWidget`), and exit it with Escape key (this part is important, you have to exit by cancelling; exiting by selecting won't show the bug).  Then hover over the same widgets.  Nothing is highlighted.  Next, click elsewhere in the dialog.  Then hover again; the widgets are now highlighted.  As I mentioned before, this bug is not new, so it has nothing to do with the current rework.



> And maybe we should **always** update the back buffer (even though we did not with the old code):
> 
> From `SDL_RenderPresent`doc:
> 
> > The backbuffer should be considered invalidated after each present; do not assume that previous contents will exist between frames. You are strongly encouraged to call SDL_RenderClear() to initialize the backbuffer before starting each new frame's drawing, even if you plan to overwrite every pixel.

Yes, that's exactly what `clear` does; calls `SDL_RenderClear()`.  That's what I was talking about when I said we should always start with a clean slate.

**Comment by sa666666** [comment]
BTW, these effects are happening for me in Linux, Mac and Windows, on three different machines (completely different hardware and drivers).  So I don't think it's just me locally; it's a problem in the code.

**Comment by thrust26** [comment]
OK, I can reproduce both now. The latter problem is already there in 6.4.  So it is not related to the new code.

I'll look into the `OptionsDialog` issue now...

**Comment by sa666666** [comment]
Commit https://github.com/stella-emu/stella/commit/1476b2a6bf7685bf1c6b846d92698536d99d742a fixes the `OptionsDialog` shifting issue for me.

**Comment by sa666666** [comment]
A separate issue; there is another `_editMode` variable in `DataGridWidget` that mirrors one in its parent class `EditableWidget`.  The compiler is complaining about this.

**Comment by thrust26** [comment]
Why only your compiler? Grrr!

**Comment by sa666666** [comment]
I use clang, and turn on almost every warning I can.  Xcode detects it too, since it uses clang behind the scenes.  I believe there's also a clang option for Visual Studio.  But I don't know how hard it is to set up, or if it's worth the effort.  In fact I believe we _shouldn't_ do it, since sometimes VS detects things that clang misses.  The more different compilers we use, the more bugs we flush out.

Besides, I don't mind telling you about these minor things.  If all I had to worry about were minor warnings like this, I'd be a very happy person :smile:

**Comment by sa666666** [comment]
There's also a bunch of warnings from `ToolTip` (unused variables), but I didn't bother mentioning this since I know you're not finished there yet :stuck_out_tongue:

**Comment by thrust26** [comment]
> In fact I believe we shouldn't do it, since sometimes VS detects things that clang misses. The more different compilers we use, the more bugs we flush out.

I have seen some methods definitions without implementations. Never bothered, maybe I should?

**Comment by sa666666** [comment]
What do you mean?  Can you give an example?

**Comment by thrust26** [comment]
If I find one... :smile:

**Comment by sa666666** [comment]
If you mean a method definition exists and no corresponding method body, then the extended tests I run pick up on that.  And then I remove them (or mark them as such).  Also, methods that are defined but never used (aka, dead code) are also picked up in these extended tests.  However, it takes over 30 minutes to run on my machine, whereas Stella compiles in ~20 seconds.  Just to illustrate how CPU-intensive this is, which is why I don't run it very often.  Usually only before a major release.

**Comment by thrust26** [comment]
So probably you removed them in between.

**Comment by thrust26** [comment]
@sa666666 Why is a surface cut when it extends the border of the application? E.g. when a tool tip is wide and would have to be displayed outside Stella's window?

**Comment by sa666666** [comment]
How else could it work?  The window boundaries are the 'entire world' to SDL. It doesn't know what exists (if anything) outside those boundaries.  There's no way for SDL to draw outside its own window.  The best we can do here is reposition the dialog so that is falls within the boundary.

Look at `RomListWidget` for an example.  Right-click to get a dialog.  Now move the mouse to bottom-right area and do it again.  The dialog is positioned such that it's still in the window.

**Comment by sa666666** [comment]
Also, some small issues with tooltips (but good first attempt!).

![Screenshot_20201116_100613](https://user-images.githubusercontent.com/16061066/99258788-a7c25880-27f3-11eb-8c68-232bc744fda6.png)

![Screenshot_20201116_100632](https://user-images.githubusercontent.com/16061066/99258789-a7c25880-27f3-11eb-9187-77bf0e618ba9.png)

I'm using HiDPI mode, and it looks like the tooltips are positioned based on absolute coordinates rather than relative.  Maybe look at the `PopupWidget`/`ContextMenu` interaction, and how it works there for when HiDPI is either on or off.  This was the way those UI elements were drawn until I took HiDPI into account.

Also, there's the issue of the font being too small, but that's an easy fix I guess.

**Comment by thrust26** [comment]
Regarding the surfaces, would it be possible to create a bottom, fully transparent, empty surface for the whole current display and then but the other surfaces on top of it? So that display content could be anywhere on the display?

Also I have to check who Stella can be dragged around, maybe we could create fully independent surfaces that way.

And yes, HiDPI and font size were not on my top priorities yet.

**Comment by sa666666** [comment]
SDL (and indeed all graphical toolkits) work on the idea of a 'window'.  Even in fullscreen, it's just a borderless window.  When a window is created in `FBBackendSDL2`, that's all we have to work with.  And TBH, that's as far as I'd want to go with this.  If I understand you correctly, you want to create a transparent window the size of the entire monitor, then have Stella be in only some small part of it?  SDL can't do that, and TBH that starts to fall into the 'window manager' territory of the OS.  So really outside the scope of SDL, and definitely of Stella.

Now, if you really want to pursue this, then I guess we can look at the multi-window functionality of SDL.  That is, you can create separate _windows_ at the OS level.  But then we run into issues on platforms that don't have a windowing system, and also what happens when you switch to fullscreen mode (where I think only the main window will be visible).

Honestly, I think the best thing to do is stick with the one window we have, and just restrict our drawing to it.  This stuff starts to get _really_ complicated if you move into multiple windows.  It breaks the current design completely, and I'm sure I don't want to go that far into it.

**Comment by sa666666** [comment]
Just to add to this, it's very likely a tooltip in other applications is a borderless window.  That's how it's able to 'move' and be drawn outside the main window; the OS handles it as a separate window, with the close button/etc removed.

I think it's much easier to just reposition the tooltip so, if it would be clipped, it is instead moved.  `PopupWidget`/`ContextMenu` does this, as does the right-click dialog in `RomListWidget`.

**Comment by thrust26** [comment]
Yes, that's probably the way to go.

**Comment by sa666666** [comment]
The latest commits remove the highlighting upon gaining/losing focus.  Is this intentional, or a WIP?

**Comment by thrust26** [comment]
No, that's a bug.

**Comment by thrust26** [comment]
I think this is done for now.
